nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -322,23 +322,42 @@ geometry_msgs::msg::PoseStamped RegulatedPurePursuitController::getLookAheadPoin
   // If the no pose is not far enough, take the last pose
   if (goal_pose_it == transformed_plan.poses.end()) {
     goal_pose_it = std::prev(transformed_plan.poses.end());
-  } else if (params_->use_interpolation && goal_pose_it != transformed_plan.poses.begin()) {
+  }
+  if (params_->use_interpolation && goal_pose_it != transformed_plan.poses.begin()) {
+    // We have more than one pose in the plan, so we can interpolate
+    auto goal_position = goal_pose_it->pose.position;
+    auto prev_pose_it = std::prev(goal_pose_it);
+    auto prev_position = prev_pose_it->pose.position;
+
+    if (goal_pose_it == std::prev(transformed_plan.poses.end())) {
+      // This means that the last pose is inside the lookahead distance
+      // Project forward based on the vector between the last two poses
+      // to create the carrot beyond the end of the plan
+      // This will ensure that the minimum lookahead distance constraint is preserved
+      // and that the gain of the steering isn't effectively increased near the goal
+      auto dx = goal_position.x - prev_position.x;
+      auto dy = goal_position.y - prev_position.y;
+
+      auto d = std::hypot(dx, dy);
+      // Projecting past the last point by the lookahead distance ensures that the point
+      // is at least the lookahead distance away
+      // (maybe more, but we will interpolate in the next step)
+      goal_position.x += dx / d * lookahead_dist;
```

Comment:
I... don't think this is correct

If the total distance between points is 1 unit and equal component X and Y, then `dx` and `dy` would be `0.70710678118` for a total `d` of `1.0`. In that case, we're adding more to the goal pose than would be the length of the `lookahead_distance`. 

I think you need to do some trig with the angles to see what component of each you need to apply like the commit I originally linked to did, unless I'm missing something.

Reply:
Yes, that is deliberate because the direction is the important bit. The distance is truncated in the following code by the call:
https://github.com/ros-planning/navigation2/blob/681d7e15acbb27618272e0530d7b8f385d614fad/nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp#L353

> I think you need to do some trig with the angles to see what component of each you need to apply like the commit I originally linked to did, unless I'm missing something.

This is the same as in the linked commit https://github.com/ros-planning/navigation2/blob/a9d8c4c9889c7d8684296c5d7733f507255520c1/nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp#L232-L239 just without using trig functions to convert to an intermediate angle.

The distance is set to the lookahead distance so that it is outside the radius and `circleSegmentationIntersection()` succeeds.

The difference is approach is that in the previous work in https://github.com/ros-planning/navigation2/commit/a9d8c4c9889c7d8684296c5d7733f507255520c1 the remaining lookahead distance is calculated by taking the "path end as carrot" distance, and working out how much shorter than the lookahead distance it is. This distance is then added to the last segment of the path and treated as the new carrot. (There is an implicit assumption that the direction from the robot to the last path pose is the same as the direction of the last segment, but I suspect that in practice this would make very little difference/be very unlikely to cause problems as it would need odd dog-leg paths at the scale of the lookahead distance for it to matter).
My approach checks to see if the last pose is being selected before the call to find the circle-segment intersection (which runs for all interpolations along the path), and shifts the goal pose before this call. So every iteration uses the circle-segment intersection approach, even when it is on the last segment.

Reply:
Trying to find a slot to test on a real big robot. Give me 24h

> (There is an implicit assumption that the direction from the robot to the last path pose is the same as the direction of the last segment, but I suspect that in practice this would make very little difference/be very unlikely to cause problems as it would need odd dog-leg paths at the scale of the lookahead distance for it to matter).

Indeed, I developed and tested https://github.com/ros-planning/navigation2/commit/a9d8c4c9889c7d8684296c5d7733f507255520c1 for straight paths only. I agree this should be mostly fine for general paths but I did not make the exercise of thinking and testing about corner cases.